qa/update-tests/src/test/java/io/camunda/zeebe/test/migration/MultiTenancyMigrationTest.java

```diff
@@ -0,0 +1,214 @@
+import io.camunda.zeebe.qa.util.actuator.PartitionsActuator;
+import io.camunda.zeebe.qa.util.testcontainers.ZeebeTestContainerDefaults;
+import io.camunda.zeebe.snapshots.impl.FileBasedSnapshotId;
+import io.camunda.zeebe.test.util.Strings;
+import io.camunda.zeebe.test.util.junit.RegressionTest;
+import io.zeebe.containers.ZeebeContainer;
+import io.zeebe.containers.ZeebeVolume;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.Nested;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+
+@Testcontainers
+final class MultiTenancyMigrationTest {
+  private static final DockerImageName PRE_MIGRATION_VERSION =
+      DockerImageName.parse("camunda/zeebe:8.2.21");
+  private static final DockerImageName POST_MIGRATION_VERSION =
+      ZeebeTestContainerDefaults.defaultTestImage();
+
+  private final ZeebeVolume volume = ZeebeVolume.newVolume();
+
+  @Container
+  private final ZeebeContainer container =
+      new ZeebeContainer(PRE_MIGRATION_VERSION)
+          .withEnv("ZEEBE_LOG_LEVEL", "debug")
+          // memory limit has to be higher initially because of the unbounded cache in process state
+          // and to make it faster to generate our large snapshot
+          .withEnv("ZEEBE_BROKER_EXPERIMENTAL_ROCKSDB_MEMORYLIMIT", "512MB")
+          .withCreateContainerCmdModifier(cmd -> constrainMemory(cmd, 2048))
+          .withEnv("ZEEBE_BROKER_EXPERIMENTAL_ENGINE_CACHES_DRGCACHECAPACITY", "5")
+          .withZeebeData(volume);
+
+  private void forceSnapshot() {
+    final var partitions = PartitionsActuator.of(container);
+    partitions.takeSnapshot();
+    Awaitility.await("Snapshot is taken on all partitions")
+        .atMost(Duration.ofSeconds(60))
+        .until(
+            () ->
+                partitions.query().values().stream()
+                    .allMatch(
+                        p ->
+                            p.snapshotId() != null
+                                && FileBasedSnapshotId.ofFileName(p.snapshotId()).isPresent()));
+  }
+
+  private void constrainMemory(final CreateContainerCmd cmd, final long memoryLimitMb) {
+    final var hostConfig = Optional.ofNullable(cmd.getHostConfig()).orElseGet(HostConfig::new);
+    cmd.withHostConfig(hostConfig.withMemory(memoryLimitMb * 1024 * 1024L)).withUser("0:0");
+  }
+
+  private void reconfigureContainerWithNewerVersion() {
+    container.stop();
+    container
+        // constrain memory to "expected" limits: 128MB * 3 + some JVM memory <= 600MB
+        // this means 600MB should be enough to run Zeebe, including performing migrations
```

Comment:
I want to talk to Lars more about the differences between DinD and normal Docker. It might be further incentive to enable some stuff on Testcontainers Cloud, depending on the reasons.